File: 2016/src/day15.go
package src

import (
    "fmt"
    "strings"
    "strconv"
)

func Solve15(input string) string {
    var ans1 int
    var ans2 int

    values := map[int]int{}

    lines := strings.Split(input, "\n")
    for i, line := range lines {
        splits := strings.FieldsFunc(line, func (r rune) bool { return r == ' ' || r == '.' })
        pos, _ := strconv.Atoi(splits[3])
        curPos, _ := strconv.Atoi(splits[11])

        values[pos] = curPos + i + 1
    }

    time := 0
    mul := 1
    for pos, curPos := range values {
        for (curPos + time) % pos != 0 {
            time += mul
        }
        mul *= pos
    }

    ans1 = time

    for (len(lines) + 1 + time) % 11 != 0 {
        time += mul
    }

    ans2 = time

    return fmt.Sprintf("%d, %d", ans1, ans2)
}

File: 2017/src/day15.js
export default function solve15(input) {
    let ans1 = 0;
    let ans2 = 0;

    let mask = Math.pow(2, 31) - 1;
    let compMask = Math.pow(2, 16) - 1;

    let lines = input.split("\n");
    let a = parseInt(lines[0].split(" ")[4]);
    let b = parseInt(lines[1].split(" ")[4]);

    for (let i = 0; i < 40000000; i++) {
        a = (a * 16807) % 2147483647;
        b = (b * 48271) % 2147483647;
        if ((a & compMask) === (b & compMask)) {
            ans1++;
        }
    }

    a = parseInt(lines[0].split(" ")[4]);
    b = parseInt(lines[1].split(" ")[4]);

    for (let i = 0; i < 5000000; i++) {
        do {
            a = (a * 16807) % 2147483647;
        } while ((a & 3) != 0);
        do {
            b = (b * 48271) % 2147483647;
        } while ((b & 7) != 0);

        if ((a & compMask) === (b & compMask)) {
            ans2++;
        }
    }

    return [ans1, ans2];
}
